_drafts/zero-to-main-2.md

```diff
@@ -0,0 +1,536 @@
+the LMA is where your JTAG loader needs to place the section and the VMA is
+where the section is found during execution.
+
+You can think of the LMA as the address "at rest" and the VMA the address during
+execution i.e. when the device is on and the program is running.
+
+The syntax to specify the LMA is relatively straightforward: every address is
+two part: <VMA> AT <LMA>. In our case it looks like this:
+
+```
+.data :
+{
+    *(.data*);
+} > ram AT > rom /* "> ram" is the VMA, "> rom" is the LMA */
+```
+
+Note that instead of appending a section to a memory region, you can explicity
+specify an address like so:
+
+```
+.data ORIGIN(ram) /* VMA */ : AT(ORIGIN(rom)) /* LMA */
+{
+    . = ALIGN(4);
+    _sdata = .;
+    *(.data*);
+    . = ALIGN(4);
+    _edata = .;
+}
+```
+
+Where `ORIGIN(<region>)` is a simple way to specify the start of a region. You
+can enter an address in hex as well.
+
+And we're done! Here's our complete linker script with every section:
+
+```
+MEMORY
+{
+  rom      (rx)  : ORIGIN = 0x00000000, LENGTH = 0x00040000
+  ram      (rwx) : ORIGIN = 0x20000000, LENGTH = 0x00008000
+}
+
+STACK_SIZE = 0x2000;
+
+/* Section Definitions */
+SECTIONS
+{
+    .text :
+    {
+        KEEP(*(.vectors .vectors.*))
+        *(.text*)
+        *(.rodata*)
+    } > rom
+
+    /* .bss section which is used for uninitialized data */
+    .bss (NOLOAD) :
+    {
+        *(.bss*)
+        *(COMMON)
+    } > ram
+
+    .data :
+    {
+        *(.data*);
+    } > ram AT >rom
+
+    /* stack section */
+    .stack (NOLOAD):
+    {
+        . = ALIGN(8);
+        . = . + STACK_SIZE;
+        . = ALIGN(8);
+    } > ram
+
+    _end = . ;
+}
+```
+
+You can find the full details on linker script sections syntax in the
+[ld
+manual](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/4/html/Using_ld_the_GNU_Linker/sections.html#OUTPUT-SECTION-DESCRIPTION).
+
+### Variables
+
+In the first post, our ResetHandler relied on seemingly magic variables to know
+the address of each of our sections of memory. It turns out, those variable came
+from the linker script!
+
+In order to make section addresses available to code, the linker is able
+too generate symbols and add them to the program. 
+
+You can find the syntax in the [linker
+documentation](https://access.redhat.com/documentation/en-US/Red_Hat_Enterprise_Linux/4/html/Using_ld_the_GNU_Linker/assignments.html),
+it looks exactly like a C assignment: `symbol = expression;`
+
+Here, we need:
+1. `_etext` the end of the code in `.text` section in flash.
+2. `_sdata` the start of the `.data` section in RAM
+3. `_edata` the end of the `.data` section in RAM
+4. `_sbss` the start of the `.bss` section in RAM
+5. `_ebss` the end of the `.bss` section in RAM
+
+They are all relatively straightforward: we can assign our symbols to the value
+of the location counter (`.`) at the start and at the end of each section
+definition.
+
+The code is below:
+```
+    .text :
```

Comment:
Not a big deal but these are a little different than what is in `samd21g18a_flash.ld` on master. Not sure if we want the final one to match that exactly ...

Reply:
I'll push an update to zero-to-main